plenum/test/restart/test_persist_pp.py

```diff
@@ -0,0 +1,134 @@
+    primary_replicas = get_primary_replicas(txnPoolNodeSet)
+
+    sdk_send_batches_of_random(looper, txnPoolNodeSet,
+                               sdk_pool_handle, sdk_wallet_client,
+                               batches_count, batches_count,
+                               timeout=tconf.Max3PCBatchWait)
+
+    seq_no_before_1 = primary_replicas[1].lastPrePrepareSeqNo
+    seq_no_before_2 = primary_replicas[2].lastPrePrepareSeqNo
+
+    # Restart backup primaries
+    restart_nodes(looper, txnPoolNodeSet, [primary_replicas[1].node, primary_replicas[2].node],
+                  tconf, tdir, allPluginsPath)
+    primary_replicas_new = get_primary_replicas(txnPoolNodeSet)
+    assert primary_replicas_new[0] == primary_replicas[0]
+
+    # Check that we've persisted last send pre-prepare on backup primary replicas
+    assert seq_no_before_1 == \
+           node_status_db_serializer.deserialize(
+               primary_replicas_new[1].node.nodeStatusDB.get(LAST_SENT_PRE_PREPARE))[2]
+    assert seq_no_before_2 == \
+           node_status_db_serializer.deserialize(
+               primary_replicas_new[2].node.nodeStatusDB.get(LAST_SENT_PRE_PREPARE))[2]
+
+
+def test_clear_persisted_last_pp_after_view_change(looper, txnPoolNodeSet, tconf, tdir,
+                                                   sdk_pool_handle, sdk_wallet_client, allPluginsPath):
+    sdk_send_batches_of_random_and_check(looper, txnPoolNodeSet, sdk_pool_handle,
+                                         sdk_wallet_client, batches_count, batches_count)
+
+    primary_replicas = get_primary_replicas(txnPoolNodeSet)
+    old_3pc = primary_replicas[0].last_ordered_3pc
+    # Restart master primary to make a view_change
+    restart_nodes(looper, txnPoolNodeSet, [primary_replicas[0].node], tconf, tdir, allPluginsPath,
+                  after_restart_timeout=tconf.ToleratePrimaryDisconnection + 1)
+    primary_replicas_new = get_primary_replicas(txnPoolNodeSet)
+    assert primary_replicas[0] not in primary_replicas_new
+
+    # Check that we've cleared last send pre-prepare on every primary replica
+    assert LAST_SENT_PRE_PREPARE not in primary_replicas_new[1].node.nodeStatusDB
+
+    sdk_send_batches_of_random_and_check(looper, txnPoolNodeSet, sdk_pool_handle,
+                                         sdk_wallet_client, batches_count, batches_count)
+    persisted = node_status_db_serializer.deserialize(
+        primary_replicas_new[1].node.nodeStatusDB.get(LAST_SENT_PRE_PREPARE))
+    assert persisted[2] == batches_count
+    assert persisted[1] == old_3pc[0] + 1
+
+
+def test_clear_persisted_last_pp_after_pool_restart(looper, txnPoolNodeSet, tconf, tdir,
+                                                    sdk_pool_handle, sdk_wallet_client):
+    sdk_send_batches_of_random_and_check(looper, txnPoolNodeSet, sdk_pool_handle,
+                                         sdk_wallet_client, batches_count, batches_count)
```

Comment:
After sending requests we must wait until the 3PC-batches are ordered by the primary replicas of all the instances (not only master).

Reply:
fixed